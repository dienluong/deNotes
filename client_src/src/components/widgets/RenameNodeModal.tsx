import React from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';

// Types
import { FormEvent, ChangeEvent } from 'react';
export type PropsT = {
  nodeType: NodeTypeT,
  currentName: string,
  onCloseHandler: () => unknown;
  onSubmitHandler: ({ name }: { name: string }) => unknown;
};

export default function RenameNodeModal({ currentName, onCloseHandler, onSubmitHandler }: PropsT) {
  const [open, setOpen] = React.useState(true);
  let textFieldValue = currentName;

  function handleClose() {
    setOpen(false);
    if (typeof onCloseHandler === 'function') {
      onCloseHandler();
    }
  }

  function handleChange(event: ChangeEvent<HTMLInputElement> ) {
    textFieldValue = event.target.value;
  }

  function submit(event: FormEvent<HTMLElement>) {
    handleClose();
    // TODO: Trim whitespaces
    if (typeof onSubmitHandler === 'function') {
      onSubmitHandler({ name: textFieldValue });
    }
    event.preventDefault();
  }

  return (
    <div>
      <Dialog
        open={ open }
        onClose={ handleClose }
        aria-labelledby="rename-node-dialog-title"
      >
        <DialogTitle id="rename-node-dialog-title">{ `Enter Name for ${currentName}` }</DialogTitle>
        <DialogContent>
          <form onSubmit={ submit } noValidate autoComplete="off">
            <TextField
              autoFocus
              margin="dense"
              id="name"
              label="Name"
              type="text"
              required
              fullWidth
              variant="outlined"
              defaultValue={ currentName }
              onChange={ handleChange }
              onFocus={ (e: React.FocusEvent<HTMLInputElement>) => e.target.select() }
            />
          </form>
        </DialogContent>
        <DialogActions>
          <Button aria-label={ 'Cancel Rename Dialog' } onClick={ handleClose } color="primary">
            Cancel
          </Button>
          <Button aria-label={ 'OK Rename Dialog' } onClick={ submit } color="primary">
            OK
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  );
}
