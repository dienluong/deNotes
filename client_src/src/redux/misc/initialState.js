import uuid from 'uuid/v1';
import Delta from 'quill-delta';
const _ID_DELIMITER = process.env.REACT_APP_ID_DELIMITER;

// TODO: Is it the best place to define this?
const _rootNode = [
  {
    title: '/',
    subtitle: '',
    uniqid: uuid(),
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    type: 'folder',
    expanded: true,
    children: [],
  },
];

const epoch = 0;
export default {
  notesTree: _rootNode,
  activeNode: {
    id: _rootNode[0].id,
    path: [_rootNode[0].id],
  },
  editorContent: {
    id: '',
    title: '',
    content: '<p><br></p>',
    delta: new Delta(),
    dateCreated: epoch,
    dateModified: epoch,
    readOnly: false,
  },
};
