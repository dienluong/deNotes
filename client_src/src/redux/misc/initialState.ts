import uuid from 'uuid/v4';
import Delta from 'quill-delta';
import { nodeTypes, DEFAULT_ID_DELIMITER} from '../../utils/appCONSTANTS';

const _ID_DELIMITER = process.env.REACT_APP_ID_DELIMITER || DEFAULT_ID_DELIMITER;

const _rootNode: TreeNodeT = {
    title: '/',
    subtitle: '',
    uniqid: uuid(),
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    type: nodeTypes.FOLDER,
    expanded: true,
    children: [],
};

const epoch = 0;
const now = Date.now();
const initialState: AppStateT = {
  userInfo: {
    id: 'default-user',
  },
  notesTree: {
    id: uuid(),
    tree: [_rootNode],
    editMode: false,
    editModeSelectedNodes: [],
    dateCreated: now,
    dateModified: now,
  },
  // activeNode contains the ID of the current active node, and its path, in the tree;
  // the active node can be an item (note) or a folder;
  // if the active node is the root (with no child selected), then ID is constant NONE_SELECTED and path is [NONE_SELECTED]
  activeNode: {
    id: _rootNode.id,
    path: [_rootNode.id],
  },
  editorContent: {
    id: 'default',
    title: '',
    content: '<p><br></p>',
    delta: new Delta(),
    dateCreated: epoch,
    dateModified: epoch,
    readOnly: true,
  },
  modalInfo: {
    type: '',
    props: {},
  }
};

export default initialState;
