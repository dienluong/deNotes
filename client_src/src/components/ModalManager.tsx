import React from 'react';
import { connect } from 'react-redux';
import * as rootReducer from '../redux/reducers';
import RenameNodeModal from './widgets/RenameNodeModal';

// Types
interface MapStatePropsT {
  type: string;
  props: Object;
}

function mapStateToProps(state: AppStateT): MapStatePropsT {
  return {
    type: rootReducer.selectModalInfoType(state),
    props: rootReducer.selectModalInfoProps(state),
  };
}

const MODAL_COMPONENTS: { [index: string]: React.ReactNode } = {
  'RENAME_NODE': RenameNodeModal,
};

const modalTypes = Object.keys(MODAL_COMPONENTS);
export const MODAL_TYPES: { [index: string]: string } = {};
modalTypes.forEach(type => MODAL_TYPES[type] = type);

function ModalManager({ type, props }: MapStatePropsT) {
  if (!type || !MODAL_COMPONENTS[type]) {
    return null;
  }
  const ChosenModal = MODAL_COMPONENTS[type];

  //@ts-ignore
  return <ChosenModal {...props} />
}

export default connect(mapStateToProps)(ModalManager);
