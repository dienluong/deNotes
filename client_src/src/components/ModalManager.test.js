import React from 'react';
import createMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import initialState from '../redux/misc/initialState';
import ModalManager, { MODAL_TYPES } from './ModalManager';
import RenameNodeModal from './widgets/RenameNodeModal';

import { createSerializer } from 'enzyme-to-json';

// Configure Enzyme
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
let wrapper = {};

// Configure snapshot serializer
expect.addSnapshotSerializer(createSerializer());

afterEach(() => {
  if (typeof wrapper.unmount === 'function') {
    wrapper.unmount();
  }
});

it('renders specified modal correctly with prop values taken from redux store', () => {
  let mockedInitialState = {
    ...initialState,
    modalInfo: {
      type: MODAL_TYPES.RENAME_NODE,
      props: {
        prop1: 'abcdef',
        prop2: 123456,
        prop3: [1, 'A', false],
      },
    },
  };
  let store = createMockStore([])(mockedInitialState);
  const expectedProps = { ...mockedInitialState.modalInfo.props };

  wrapper = shallow(<Provider store={ store }><ModalManager /></Provider>).dive({ context: { store } }).dive({ context: { store } });
  expect(wrapper).toMatchSnapshot();

  expect(wrapper.find(RenameNodeModal).exists()).toBe(true);
  expect(wrapper.find(RenameNodeModal).props()).toMatchObject(expectedProps);
});

it('renders nothing if give invalid modal type', () => {
  let mockedInitialState = {
    ...initialState,
    modalInfo: {
      type: MODAL_TYPES['BOGUS_MODAL'],
      props: {
        prop1: 'abcdef',
        prop2: 123456,
        prop3: [1, 'A', false],
      },
    },
  };
  let store = createMockStore([])(mockedInitialState);

  wrapper = shallow(<Provider store={ store }><ModalManager /></Provider>).dive({ context: { store } }).dive({ context: { store } });
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.isEmptyRender()).toBe(true);

  wrapper.unmount();

  mockedInitialState = {
    ...initialState,
    modalInfo: {
      props: {
        prop1: 'abcdef',
        prop2: 123456,
        prop3: [1, 'A', false],
      },
    },
  };
  store = createMockStore([])(mockedInitialState);

  wrapper = shallow(<Provider store={ store }><ModalManager /></Provider>).dive({ context: { store } }).dive({ context: { store } });
  expect(wrapper.isEmptyRender()).toBe(true);
});
