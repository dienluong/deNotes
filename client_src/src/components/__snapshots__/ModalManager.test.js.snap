// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders nothing if give invalid modal type 1`] = `""`;

exports[`renders specified modal correctly with prop values taken from redux store 1`] = `
<RenameNodeModal
  prop1="abcdef"
  prop2={123456}
  prop3={
    Array [
      1,
      "A",
      false,
    ]
  }
/>
`;
